src/justenoughharfbuzz.c

```diff
@@ -278,44 +257,110 @@ int shape (lua_State *L) {
     }
     /* Cleanup */
     hb_buffer_destroy(buf);
-    hb_font_destroy(hbFont);
-    hb_face_destroy(hbFace);
-    hb_blob_destroy(blob);
 
     free(features);
     return glyph_count;
 }
 
-int get_glyph_dimensions(lua_State *L) {
-  size_t font_l;
-  const char * font_s = luaL_checklstring(L, 1, &font_l);
-  unsigned int font_index = (unsigned int)luaL_checknumber(L, 2);
-  double point_size = (unsigned int)luaL_checknumber(L, 3);
-  hb_codepoint_t glyphId = (hb_codepoint_t)luaL_checknumber(L, 4);
-
-  hb_blob_t* blob = hb_blob_create(font_s, font_l, HB_MEMORY_MODE_WRITABLE,
-      (void*)font_s, NULL);
-  hb_face_t* hbFace = hb_face_create(blob, font_index);
-  hb_font_t* hbFont = hb_font_create(hbFace);
-  unsigned int upem = hb_face_get_upem(hbFace);
-  hb_font_set_scale(hbFont, upem, upem);
-
-  hb_variation_t opsz = { HB_TAG('o', 'p', 's', 'z'), point_size };
-  hb_font_set_variations(hbFont, &opsz, 1);
-
-  if (can_use_ot_funcs(hbFace)) {
-    hb_ot_font_set_funcs(hbFont);
-  } else {
-    /*
-      Note that using FT may cause differing vertical metrics for CFF fonts.
-      SILE will give a one-time warning if this is the case.
-    */
-    hb_ft_font_set_funcs(hbFont);
+int instanciate(lua_State *L) {
+  unsigned int data_l = 0;
+  const char * data_s = NULL;
+#ifdef HAVE_HARFBUZZ_SUBSET
+
+  hb_font_t* font = get_hb_font(L, 1);
+  hb_face_t* face = hb_font_get_face(font);
+
+  if (hb_ot_var_has_data(face) &&
+      /* hb-subset does not support instanciating CFF2 table yet */
+      !has_table(face, HB_TAG('C','F','F','2'))) {
+    hb_subset_input_t * input;
+
+    input = hb_subset_input_create_or_fail();
+    if (input) {
+      hb_ot_var_axis_info_t* axes;
+      unsigned int nAxes;
+      unsigned int nCoords;
+      const float* coords;
+      hb_face_t* subset;
+      hb_set_t* glyphs;
+      hb_set_t* tables;
+
+      hb_subset_input_set_flags(input,
+                                HB_SUBSET_FLAGS_RETAIN_GIDS |
+                                HB_SUBSET_FLAGS_NAME_LEGACY |
+                                HB_SUBSET_FLAGS_GLYPH_NAMES |
+                                HB_SUBSET_FLAGS_NO_PRUNE_UNICODE_RANGES);
+
+      /* Keep all glyphs */
+      glyphs = hb_subset_input_set(input, HB_SUBSET_SETS_GLYPH_INDEX);
+      hb_set_invert(glyphs);
```

Comment:
Obviously not required for this, but one wonders if we couldn't get smaller PDF files by also subsetting fonts inside SILE at the end of document generation, now that we optionally include the subsetter.

I've had some issues before submitting documents to Unicode Consortium via email due to their size and used various hacks to get them down to size. I think that in a font specimen with many different instances all in use at once it'd become more important, but ultimately for this PR is premature optimization. :)

Reply:
libtexpdf do subset the fonts somehow, it clears up the outlines of unused glyphs but keeps the glyphs IDs (it is is probably a bit less efficient file size-wise than complete subsetting, but probably not by much). We can’t subset here because we share the instantiated file across all font invocations.